compiler-core/src/diagnostic.rs

```diff
@@ -19,18 +21,37 @@ pub struct Label {
     pub span: SrcSpan,
 }
 
+trait ToCodespanLabel {
+    fn to_codespan_label(&self, fileid: usize) -> CodespanLabel<usize>;
+}
+
+impl ToCodespanLabel for Label {
+    fn to_codespan_label(&self, fileid: usize) -> CodespanLabel<usize> {
+        let label = CodespanLabel::new(
+            LabelStyle::Primary,
+            fileid,
+            (self.span.start as usize)..(self.span.end as usize),
+        );
+        match &self.text {
+            None => label,
+            Some(text) => label.with_message(text.clone()),
+        }
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Location {
+pub struct ExtraLocation {
```

Comment:
It looks like the vast majority of the time the path is the same, so I think location information should be optional.